src/EVM.hs

```diff
@@ -3004,16 +3012,34 @@ instance VMOps Symbolic where
       choosePath loc Unknown =
         choose . PleaseChoosePath condSimp $ choosePath loc . Case
 
-  oneSolution ewordExpr continue = do
+  -- numBytes allows us to specify how many bytes of the returned value is relevant
+  -- if it's e.g.a JUMP, only 2 bytes can be relevant. This allows us to avoid
+  -- getting solutions that are nonsensical
+  manySolutions ewordExpr numBytes continue = do
     pathconds <- use #constraints
-    query $ PleaseGetSol ewordExpr pathconds $ \case
-      Just concVal -> do
+    query $ PleaseGetSols ewordExpr numBytes pathconds $ \case
+      Just concVals -> do
         assign #result Nothing
-        pushTo #constraints $ Expr.simplifyProp (ewordExpr .== (Lit concVal))
-        continue $ Just concVal
+        case (length concVals) of
+          0 -> continue Nothing
+          1 -> runOne $ head concVals
+          _ -> choose . PleaseChoosePath ewordExpr $ runMore concVals
       Nothing -> do
         assign #result Nothing
         continue Nothing
+    where
+      runMore vals leftOrRight = do
+        case length vals of
+          -- if 2, we run both, otherwise, we run 1st and run ourselves with the rest
+          2 -> if leftOrRight then runOne $ head vals
+               else runOne (head $ tail vals)
+          _ -> if leftOrRight then runOne $ head vals
+               else choose . PleaseChoosePath ewordExpr $ runMore (tail vals)
```

Comment:
Can you explain this a bit? I just probably don't understand `choose` and `PleaseChoosePath`, but what does the parameter `leftOrRight` represents and where is it actually passed in?

For the case of `2`, if `leftOrRight` is `false`, isn't this just ignoring the first value? I don't see where the "we run both" from the comment actually happens.

Reply:
Yeah, sorry. It's actually not CHOOSING anything. It's running both. It USED to be choosing things, because there was a debugger! But no longer. Let me rename that thing and then it'll be lot easier to understand :)

Reply:
Yay, OK, it should be a lot cleaner now, renamed all relevant functions and datatypes. Can you re-review?